src/resource_clients/request_queue.ts

```diff
@@ -96,6 +97,102 @@ export class RequestQueueClient extends ResourceClient {
         return cast(parseDateFields(pluckData(response.data)));
     }
 
+    protected async _batchAddRequestsWithRetries(
+        requests: Array<Omit<RequestQueueClientRequestSchema, 'id'>>,
+        options: RequestQueueClientBatchAddRequestOptions = {},
+    ): Promise<RequestQueueClientBatchAddRequestsResult> {
+        ow(requests, ow.array.ofType(ow.object.partialShape({
+            id: ow.undefined,
+        })).minLength(1).maxLength(25));
+
+        ow(options, ow.object.partialShape({
+            forefront: ow.optional.boolean,
+        }));
+
+        // Keep track of the requests that remain to be processed (in parameter format)
+        let remainingRequests = requests;
+        // Keep track of the requests that have been processed (in api format)
+        let processedRequests: Array<ProcessedRequest> = [];
+        // The requests we have not been able to process in the last call
+        // ie. those we have not been able to process at all
+        let unprocessedRequests: Array<UnprocessedRequest> = [];
+        for (let i = 0; i < 1 + this.httpClient.maxRetries; i++) {
+            const response = await this.httpClient.call({
+                url: this._url('requests/batch'),
+                method: 'POST',
+                data: remainingRequests,
+                params: this._params({
+                    forefront: options.forefront,
+                    clientKey: this.clientKey,
+                }),
+            });
+            const data = response.data.data as RequestQueueClientBatchAddRequestsResult;
+            processedRequests = [...processedRequests, ...data.processedRequests];
+            unprocessedRequests = data.unprocessedRequests;
+
+            // Get unique keys of all requests processed so far
+            const processedRequestsUniqueKeys = processedRequests.map(({ uniqueKey }) => uniqueKey);
+            // Requests remaining to be processed are the all that remain
+            // @ts-expect-error Undefined id - it seems to be added by ow validation
+            remainingRequests = requests.filter(({ uniqueKey }) => !processedRequestsUniqueKeys.includes(uniqueKey));
```

Comment:
What's the difference between `unprocessedRequests` and `remainingRequests`? I thought the unprocessed are being retried but now I see it's the diff between remaining and processed.

Reply:
The `unprocessedRequests` are the ones returned from API. But they do not contain all the fields of request, just `uniqueKey`, `method` and `url`.
`remainingRequests` are in the end the same requests, but with all the fields. It just felt more natural to me to see them as the difference between all requests (`requests`) all requests processed so far (`processedRequests`) instead of just unprocesed items from a single API call.